Android/src/org/droidplanner/android/maps/providers/baidu_map/BaiduMapOfflineMapPreference.java

```diff
@@ -0,0 +1,354 @@
+import android.widget.ListView;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import com.baidu.mapapi.map.offline.MKOLSearchRecord;
+import com.baidu.mapapi.map.offline.MKOLUpdateElement;
+import com.baidu.mapapi.map.offline.MKOfflineMap;
+import com.baidu.mapapi.map.offline.MKOfflineMapListener;
+
+import java.util.ArrayList;
+
+import org.droidplanner.android.R;
+
+public class BaiduMapOfflineMapPreference extends DialogPreference implements MKOfflineMapListener {
+    private MKOfflineMap mOffline = null;
+    private View mBindDialogView = null;
+    private TextView mCidView = null;
+    private TextView mStateView =null;
+    private EditText mCityNameView = null;
+    private ArrayList<MKOLUpdateElement> mLocalMapList = null; // downloaded offline map list
+    private LocalMapAdapter mLAdapter = null;
+
+    public BaiduMapOfflineMapPreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        setDialogLayoutResource(R.layout.fragment_baidumap_offline_map_preference);
+    }
+
+    private void initView(View view) {
+
+        mCidView = (TextView) view.findViewById(org.droidplanner.android.R.id.cityid);
+        mCityNameView = (EditText) view.findViewById(org.droidplanner.android.R.id.city);
+        mStateView = (TextView) view.findViewById(org.droidplanner.android.R.id.state);
+
+        ListView hotCityList = (ListView) view.findViewById(org.droidplanner.android.R.id.hotcitylist);
+        ArrayList<String> hotCities = new ArrayList<String>();
+        // get hot city list
+        ArrayList<MKOLSearchRecord> records1 = mOffline.getHotCityList();
+        if (records1 != null) {
+            for (MKOLSearchRecord r : records1) {
+                hotCities.add(r.cityName + "(" + r.cityID + ")" + "   --" + this.formatDataSize(r.size));
+            }
+        }
+        ListAdapter hAdapter = (ListAdapter) new ArrayAdapter<String>(this.getContext(),
+                android.R.layout.simple_list_item_1, hotCities);
+        hotCityList.setAdapter(hAdapter);
+
+        ListView allCityList = (ListView) view.findViewById(R.id.allcitylist);
+        // get all offline city map
+        ArrayList<String> allCities = new ArrayList<String>();
+        ArrayList<MKOLSearchRecord> records2 = mOffline.getOfflineCityList();
+        if (records1 != null) {
+            for (MKOLSearchRecord r : records2) {
+                allCities.add(r.cityName + "(" + r.cityID + ")" + "   --" + this.formatDataSize(r.size));
+            }
+        }
+        ListAdapter aAdapter = (ListAdapter) new ArrayAdapter<String>(this.getContext(),
+                android.R.layout.simple_list_item_1, allCities);
+        allCityList.setAdapter(aAdapter);
+
+        LinearLayout cl = (LinearLayout) view.findViewById(R.id.citylist_layout);
+        LinearLayout lm = (LinearLayout) view.findViewById(R.id.localmap_layout);
+        lm.setVisibility(View.GONE);
+        cl.setVisibility(View.VISIBLE);
```

Comment:
@droneboost you're reusing these views below, so store them as instance fields, since `findViewById(...)` is an expensive method.